File: ch9/GaussianElimination.ml
open Printf

let gaussian_elimination n aug =
  for i = 0 to n - 1 do                 (* the forward elimination phase *)
    let l = ref i in
    for j = i + 1 to n - 1 do      (* which row has largest column value *)
      if abs_float aug.(j).(i) > abs_float aug.(!l).(i) then
        l := j;                                   (* remember this row l *)
    done;
    (* swap this pivot row, reason: minimize floating point error *)
    let t = aug.(i) in aug.(i) <- aug.(!l); aug.(!l) <- t;
    for j = i + 1 to n - 1 do    (* the actual forward elimination phase *)
      for k = n downto i do
        aug.(j).(k) <- aug.(j).(k) -. aug.(i).(k) *. aug.(j).(i) /. aug.(i).(i)
      done
    done
  done;

  let ans = Array.create_float n in       (* the back substitution phase *)
  for j = n - 1 downto 0 do                           (* start from back *)
    let t = ref 0. in
    for k = j + 1 to n - 1 do
      t := !t +. aug.(j).(k) *. ans.(k)
    done;
    ans.(j) <- (aug.(j).(n) -. !t) /. aug.(j).(j)  (* the answer is here *)
  done;
  ans

let () =
  let aug = [|
    [|1.; 1.; 2.; 9.|];
    [|2.; 4.; -3.; 1.|];
    [|3.; 6.; -5.; 0.|];
  |] in
  let x = gaussian_elimination 3 aug in
  printf "X = %.1f, Y = %.1f, Z = %.1f\n" x.(0) x.(1) x.(2)

File: ch9/GaussianElimination.py
def GaussianElimination(N, mat):
  for i in range(N-1):
    l = i
    for j in range(i+1, N):
      if abs(mat[j][i]) > abs(mat[l][i]):
        l = j
    for k in range(i, N+1):
      mat[i][k], mat[l][k] = mat[l][k], mat[i][k]
    for j in range(i+1, N):
      for k in range(N, i-1, -1):
        mat[j][k] -= mat[i][k] * mat[j][i] / mat[i][i]

    ans = [0] * N
  for j in range(N-1, -1, -1):
    t = 0.0
    for k in range(j+1, N):
      t += mat[j][k] * ans[k]
    ans[j] = (mat[j][N]-t) / mat[j][j]

  return ans


def main():
  mat = [None] * 3
  mat[0] = [1, 1, 2, 9]
  mat[1] = [2, 4, -3, 1]
  mat[2] = [3, 6, -5, 0]

  X = GaussianElimination(3, mat)
  print('X = %.1lf, Y = %.1lf, Z = %.1lf' % (X[0], X[1], X[2]))


main()
